File: lib/StringList.cpp
#include <stdio.h>
#include <string>

using namespace std;

namespace ece309 {

class List {
private:
	class ListNode {
	private:
		string item; 
		ListNode *next;
	public:
		ListNode(string a) 
			{ item = a; next=NULL; }
		ListNode* getNext() { return next; }
		void setNext(ListNode *n) { next = n; }
		string getItem() { return item; }
	};
	ListNode *head;
	ListNode *tail;
public:
	List()
	{
		head = NULL;
		tail = NULL;
	}
	void push_back(string a)
	{
		ListNode *node = new ListNode(a);
		if (head==NULL) {  
			head = node;
			tail = node;
		} else {
			tail->setNext(node); 
			tail = node;         
		}
	}  
	bool remove_front(string &a)
	{
		if (!empty()) 
		{
			string copy = head->getItem(); 
			ListNode *tmp = head->getNext();
			delete head; 
			head = tmp;  
			if (tmp==NULL) 
				tail = NULL;
			return true;
		}
		return false; 
	}   
	bool empty()
	{
		return head==NULL;
	}
	string get(int n)
	{
		int i = 0;
		ListNode* currNode = head;
		
		while (currNode != NULL)
		{
			if (i == n)
				return currNode->getItem();		
			i++;
			currNode = currNode->getNext();
		}
		
		return "";
	}
	int length()
	{
		int i = 0;
		ListNode* currNode = head;
		
		while (currNode != NULL)
		{
			++i;
			currNode = currNode->getNext();
		}
		return i;
	}
	~List()
	{
		string t;
		while(!empty()) 
			remove_front(t);   
	}
};

}

File: python/StringList.py
from __future__ import annotations

class SList:
    def __init__(self):
        self.head = None
        self.tail = None
    def push_back(a: str) -> None:
        node = SListNode(a)
        if head == None:
            self.head = node
            self.tail = node
        else:
            self.tail.setNext(node)
            self.tail = node
    def remove_front(self, a: str) -> bool:
        if not self.empty():
            copy = self.head.getItem()
            tmp = self.head.getNext()
            self.head = tmp
            if tmp == None:
                self.tail = None
            return True
        return False
    def empty(self) -> bool:
        return self.head == None
    def get(self, n: int) -> str:
        i = 0
        currNode = self.head
        while currNode != None:
            if i == n:
                return currNode.getItem()
            i += 1
            currNode = currNode.getNext()
        return ""
    def length(self) -> int:
        i = 0
        currNode = self.head
        while currNode != None:
            i += 1
            currNode = currNode.getNext()
        return i
    def __del__(self):
        t = "";
        while not self.empty():
            self.remove_front(t)

class SListNode(SList):
    def __init__(self, a: str):
        self.item = a
        self.nxt = None
    def getNext(self) -> SListNode:
        return self.nxt
    def setNext(self, n: SListNode) -> None:
        self.nxt = n
    def getItem(self) -> str:
        return item
